runtime/vm/UpcallVMHelpers.cpp

```diff
@@ -518,35 +518,20 @@ convertUpcallReturnValue(J9UpcallMetaData *data, U_8 returnType, U_64 *returnSto
 #endif /* !defined(J9VM_ENV_LITTLE_ENDIAN) */
 		break;
 	}
+#if JAVA_SPEC_VERSION == 17
 	case J9NtcPointer:
 	{
-		j9object_t retMemObject = (j9object_t)*returnStorage;
-		if (validateNonNullRetMemObject(data, retMemObject)) {
-			/* Get the native address from the returned MemorySegment(JDK20+)/MemoryAddress(JDK17/19) object
-			 * by performing a call-in so as to capture the exception for the heap address.
-			 */
-#if JAVA_SPEC_VERSION >= 20
-			*returnStorage = performCallInForRetMemObject(data, retMemObject,
-					(UDATA)J9VMOPENJ9INTERNALFOREIGNABIUPCALLMHMETADATA_GETNATIVEARGRETSEGMENTOFPTR_METHOD(data->vm));
-#else /* JAVA_SPEC_VERSION >= 20 */
-			*returnStorage = performCallInForRetMemObject(data, retMemObject,
-					(UDATA)J9VMOPENJ9INTERNALFOREIGNABIUPCALLMHMETADATA_GETNATIVEARGRETADDROFPTR_METHOD(data->vm));
-#endif /* JAVA_SPEC_VERSION >= 20 */
-		}
+		j9object_t memAddrObject = (j9object_t)*returnStorage;
+		*returnStorage = (U_64)getNativeAddrFromMemAddressObject(data, memAddrObject);
 		break;
 	}
+#else /* JAVA_SPEC_VERSION == 17 */
+	case J9NtcPointer: /* Fall through */
+#endif /* JAVA_SPEC_VERSION == 17 */
 	case J9NtcStruct:
 	{
 		j9object_t memSegmtObject = (j9object_t)*returnStorage;
-		if (validateNonNullRetMemObject(data, memSegmtObject)) {
-			/* Get the native address from the returned MemorySegment object for struct by performing a call-in.
-			 * Note:
-			 * A heap segment must be converted to a native segment in java to guarantee that all values
-			 * of the segment are safely accessed in native.
-			 */
-			*returnStorage = performCallInForRetMemObject(data, memSegmtObject,
-					(UDATA)J9VMOPENJ9INTERNALFOREIGNABIUPCALLMHMETADATA_GETNATIVEARGRETSEGMENT_METHOD(data->vm));
-		}
+		*returnStorage = (U_64)getNativeAddrFromMemSegmentObject(data, memSegmtObject);
 		break;
 	}
 	default: /* J9NtcVoid */
```

Comment:
I think this is missing a case for `J9NtcDouble`, or the comment is misleading.

Reply:
Modified the comment given there is no need to address void/long/double for the return value from upcall.